### PR TITLE
add initial block pool

### PR DESCRIPTION
**Pull request by arnetheduck**
* implement in-memory block graph
* store tail block in database
* resolve unknown parents by syncing them from peers
* introduce concept of resolved blocks and attestations - those that
follow minimal protocol rules
* update state head lazily
* log more stuff
* shortHash -> shortLog
* start 9/10 beacon nodes by default, last can be started manually
* see also #134

### CODE REVIEW

File: tests/test_block_pool.nim [hunk 1]
```diff
@@ -0,0 +1,89 @@
+# beacon_chain
+# Copyright (c) 2018 Status Research & Development GmbH
```
Comment (tersec) [inline tests/test_block_pool.nim]: Here and elsewhere, what's the best thing to do with copyright years?

Comment (Swader) [reply tests/test_block_pool.nim]: Omit or say "Copyright (c) 21st century, Status..."

I'm only half kidding, literally no one takes that in any way seriously, not even copyright lawyers (verified by two).

File: beacon_chain/spec/validator.nim [hunk 2]
```diff
@@ -269,6 +269,7 @@ func get_beacon_proposer_index*(state: BeaconState, slot: Slot): ValidatorIndex
   # TODO this index is invalid outside of the block state transition function
   #      because presently, `state.slot += 1` happens before this function
   #      is called - see also testutil.getNextBeaconProposerIndex
+  # TODO is the above still true? the shuffling has changed since it was written
```
Comment (tersec) [inline beacon_chain/spec/validator.nim]: The shuffling is still a function of `seed`, which is a function of `epoch`, so one would have to check beginning/end of epoch boundary conditions.

File: tests/test_block_pool.nim (hunk 1 above)
Comment (mratsim) [reply tests/test_block_pool.nim]: Copyright 2018-Present

File: beacon_chain/attestation_pool.nim [hunk 3]
```diff
@@ -256,18 +275,43 @@ proc add*(pool: var AttestationPool,
 
       if not found:
         a.validations.add(validation)
+        info "Attestation resolved",
+          slot = humaneSlotNum(attestation.data.slot),
+          shard = attestation.data.shard,
+          beaconBlockRoot = shortLog(attestation.data.beacon_block_root),
```
Comment (mratsim) [inline beacon_chain/attestation_pool.nim]: I suppose shortLog is a truncated print from chronicles?

File: beacon_chain/attestation_pool.nim [hunk 4]
```diff
@@ -227,8 +237,17 @@ proc add*(pool: var AttestationPool,
       pool.slots.popFirst()
       pool.startingSlot += 1
 
+  int(attestationSlot - pool.startingSlot)
+
+proc add*(pool: var AttestationPool,
+          state: BeaconState,
+          attestation: Attestation) =
+  if not validate(state, attestation, {skipValidation}): return
```
Comment (mratsim) [inline beacon_chain/attestation_pool.nim]: I think the return should be next line, it's easier to read control flow and not miss an early return that way as the line is long

File: beacon_chain/attestation_pool.nim [hunk 5]
```diff
@@ -256,18 +275,43 @@ proc add*(pool: var AttestationPool,
...
+        justifiedEpoch = humaneEpochNum(attestation.data.justified_epoch),
+        justifiedBlockRoot = shortLog(attestation.data.justified_block_root),
+        signature = shortLog(attestation.aggregate_signature),
+        validations = 1
+
+    else:
+      pool.unresolved[attestation.data.beacon_block_root] =
+        UnresolvedAttestation(
+          attestation: attestation,
+        )
```
Comment (mratsim) [inline beacon_chain/attestation_pool.nim]: TODO note that have a more efficient datastructure maybe (or open an issue after), it seems like O(n^2)

File: beacon_chain/spec/digest.nim [hunk 6]
```diff
@@ -31,6 +31,9 @@ type
   Eth2Digest* = MDigest[32 * 8] ## `hash32` from spec
   Eth2Hash* = blake2_512 ## Context for hash function
 
+func shortLog*(x: Eth2Digest): string =
```
Comment (mratsim) [inline beacon_chain/spec/digest.nim]: shortHash?

File: tests/simulation/start.sh [hunk 7]
```diff
@@ -59,35 +53,19 @@ USE_MULTITAIL="${USE_MULTITAIL:-no}" # make it an opt-in
 type "$MULTITAIL" &>/dev/null || USE_MULTITAIL="no"
 COMMANDS=()
 
-for i in $(seq 0 9); do
+for i in $(seq 0 8); do
```
Comment (zah) [inline tests/simulation/start.sh]: You explained this on the call, but the audio interrupted briefly for me. Why are we skipping the last node?

File: tests/test_block_pool.nim (hunk 1 above)
Comment (arnetheduck) [reply tests/test_block_pool.nim]: ok, let's fix this on the whole repo after merging this, in a separate PR?

File: beacon_chain/spec/validator.nim (hunk 2 above)
Comment (arnetheduck) [reply beacon_chain/spec/validator.nim]: think what would be useful here is a unit test that shows what's safe to do in terms of state-slot vs passed-in slot

File: beacon_chain/attestation_pool.nim (hunk 3 above)
Comment (arnetheduck) [reply beacon_chain/attestation_pool.nim]: yeah, the idea being that we can define a function with one name for all types instead of a different name for ever  type

File: beacon_chain/spec/digest.nim (hunk 6 above)
Comment (arnetheduck) [reply beacon_chain/spec/digest.nim]: one name for all

File: beacon_chain/beacon_node.nim [hunk 8]
```diff
@@ -3,38 +3,36 @@ import
...
   clientId = "nimbus beacon node " & version
 
   topicBeaconBlocks = "ethereum/2.1/beacon_chain/blocks"
   topicAttestations = "ethereum/2.1/beacon_chain/attestations"
-
-  stateStoragePeriod = SLOTS_PER_EPOCH.uint64 * 10 # Save states once per this number of slots. TODO: Find a good number.
-
-func shortHash(x: auto): string =
-  ($x)[0..7]
+  topicfetchBlocks = "ethereum/2.1/beacon_chain/fetch"
```
[the remark this answers is not in the document]
Comment (arnetheduck) [reply beacon_chain/beacon_node.nim]: yeah, just an example to see if it works

File: tests/simulation/start.sh (hunk 7 above)
Comment (arnetheduck) [reply tests/simulation/start.sh]: the idea is to have one node to start easily separately (using `./run_node.sh 9`)

File: tests/test_block_pool.nim (hunk 1 above)
Comment (tersec) [reply tests/test_block_pool.nim]: Yes, this PR isn't the place to fix that.

File: beacon_chain/spec/validator.nim (hunk 2 above)
Comment (tersec) [reply beacon_chain/spec/validator.nim]: Agree.

Comment (arnetheduck) [reply beacon_chain/spec/validator.nim]: https://github.com/status-im/nim-beacon-chain/issues/143
